### PR TITLE
Allow \n and # in Ok events

### LINKED ISSUE #710: Ok - '#' and "\n" should be allowed
'#' and "\n" were forbidden due to the TAP standard... But the event system in Test2 is supposed to be decoupled from TAP, so there is no good reason to disallow them at the Test2::Event::Ok level.

Both should be allowed, and Test2::Formatter::TAP (and Test::Builder::Formatter) should decide what to do with them as far as TAP goes.

Ideas for formatters:

For starters the Test::Builder::Formatter MUST preserve backcompat, so it should do what Test::Builder has always done and escape '#' with a single escape, and turn extra lines from a test description into comments.

What do we want to do with Test2::Formatter::Tap?

* Replace newlines with space?
* Make extra lines in description into comments?
 * Indent them by 5/9 characters? (length("ok - " or "not ok - "))
* Strip '#'?
* Replace '#' with space?
* Replace '#' with a unicode variant?
 * Only in utf8 mode?
* Escape '#' with a slash?
* Escape as '\\\#' (@leont?)
* Other?

### PR DESCRIPTION
**Pull request by exodist**
Fixes #710

Formatter should handle escaping these, decouple from TAP!

### CODE REVIEW

**Comment by exodist** [comment]
Performance testing reveals that there is no significant performance changes in either Test2 or Test::Builder on Test2 when using this. (WooHoo!)

File: lib/Test2/Formatter/TAP.pm [hunk 1]
```diff
@@ -154,12 +154,35 @@ sub event_ok {
     $out .= "not " unless $e->{pass};
     $out .= "ok";
     $out .= " $num" if defined($num);
+
+    # The regex form is ~250ms, the index form is ~50ms
+    my @extra;
+    defined($name) && (
```
Comment (exodist) [inline lib/Test2/Formatter/TAP.pm]: I know this is ugly as hell, and hard to read, but the performance of doing it this way vs blindly applying the regexes is unbelievable. Using && and , and similar to make this all 1 expression is also significantly faster. Normally I would not advocate for this kind of micro-optimization, but this gets done for every single OK event rendered to TAP, even tiny fractions of a second add up over a large test suite.

**Comment by exodist** [comment]
Installing Moose before this PR:

> Files=473, Tests=17304, 83 wallclock secs ( 2.43 usr  0.47 sys + 77.10 cusr  4.13 csys = 84.13 CPU)
> prove -Ilib -r t  79.60s user 4.61s system 101% cpu 1:22.87 total
> 

Installing Moose after this PR:

> Files=473, Tests=17304, 82 wallclock secs ( 2.37 usr  0.51 sys + 77.28 cusr  4.00 csys = 84.16 CPU)
> prove -Ilib -r t  79.73s user 4.52s system 101% cpu 1:22.87 total
> 

Little if any performance impact. total time is identical by chance, other numbers only slightly different.

**Comment by exodist** [comment]
Running downstream tests, assuming those are all good I will merge this and release a trial.

**Comment by exodist** [comment]
Downstream testing reveals that this breaks:

- Test::LongString
- Test::WWW::Mechanize

Together these cause a cascade of failures.

I can avoid breaking these by reverting the Test::Builder fix and keeping the Test2 fix...

**Comment by exodist** [comment]
@petdance, @leont ^

**Comment by Leont** [comment]
> Downstream testing reveals that this breaks:
> * Test::LongString*
> * Test::WWW::Mechanize

They match TAP too literally and don't expect `\` to be `\\`?

**Comment by exodist** [comment]
yes. and my guess is that they are not the only ones, just the only ones I found so far.

**Comment by exodist** [comment]
Updated the PR, it will escape both \ and # if there are # symbols in the name. If there are no # symbols it will not escape anything. I think this is the best balance between moving forward and not breaking things. Going to test downstream stuff again.

**Comment by exodist** [comment]
The update passes my downstream checks. I will cut a trial release.

**Comment by Leont** [comment]
> yes. and my guess is that they are not the only ones, just the only ones I found so far.

That's most unfortunate… :-/

> maybe if we only escape when a # is present. right now the PR always escapes # and \ maybe it should only escape them when # is present. this will still fix the harness problem, and still escape both, but will not effect tests that do literal string comparisons that only have \ in them.

I can think of at least one scenario where this can still go wrong in Test::Harness: a test description that ends with a backslash

**Comment by exodist** [comment]
I suspect that is a very uncommon this, and even more uncommon in things doing string comparisons on TAP streams, maybe I can add a special case to escape slashes if the first line of the name ends with a slash?

**Comment by exodist** [comment]
@Leont see #712

**Pull request merged**